File: example6/dataProcessor/dataProcessor.py
class DataProcessor:
    def process_sales_data(self, sales_data):
        # Duplizierter Validierungscode
        if not sales_data:
            raise ValueError("Sales data cannot be empty")
        
        if not isinstance(sales_data, list):
            raise TypeError("Sales data must be a list")
        
        processed_sales = []
        for sale in sales_data:
            # Duplizierte Validierungslogik
            if 'amount' not in sale:
                raise ValueError(f"Invalid sale record: {sale}")
            
            if 'date' not in sale:
                raise ValueError(f"Invalid sale record: {sale}")
            
            # Komplexe Verarbeitungslogik
            processed_sale = {
                'amount': float(sale['amount']),
                'date': self.normalize_date(sale['date']),
                'is_valid': self.validate_sale(sale)
            }
            processed_sales.append(processed_sale)
        
        return processed_sales
    
    def process_purchase_data(self, purchase_data):
        # Fast identischer Code wie bei Sales Data
        if not purchase_data:
            raise ValueError("Purchase data cannot be empty")
        
        if not isinstance(purchase_data, list):
            raise TypeError("Purchase data must be a list")
        
        processed_purchases = []
        for purchase in purchase_data:
            # Fast gleiche Validierungslogik
            if 'amount' not in purchase:
                raise ValueError(f"Invalid purchase record: {purchase}")
            
            if 'date' not in purchase:
                raise ValueError(f"Invalid purchase record: {purchase}")
            
            # Ähnliche Verarbeitungslogik
            processed_purchase = {
                'amount': float(purchase['amount']),
                'date': self.normalize_date(purchase['date']),
                'is_valid': self.validate_purchase(purchase)
            }
            processed_purchases.append(processed_purchase)
        
        return processed_purchases
    
    def normalize_date(self, date):
        # Datumsformatierungslogik
        try:
            # Verschiedene Datumsformatierungen normalisieren
            return str(date)
        except Exception as e:
            raise ValueError(f"Invalid date format: {date}")
    
    def validate_sale(self, sale):
        # Verkaufsvalidierung
        return (
            sale.get('amount', 0) > 0 and 
            sale.get('date') is not None
        )
    
    def validate_purchase(self, purchase):
        # Kaufvalidierung
        return (
            purchase.get('amount', 0) > 0 and 
            purchase.get('date') is not None
        )

def main():
    processor = DataProcessor()
    
    sales_data = [
        {'amount': 100, 'date': '2023-01-01'},
        {'amount': 200, 'date': '2023-02-01'}
    ]
    
    purchase_data = [
        {'amount': 150, 'date': '2023-03-01'},
        {'amount': 250, 'date': '2023-04-01'}
    ]
    
    processed_sales = processor.process_sales_data(sales_data)
    processed_purchases = processor.process_purchase_data(purchase_data)
    
    print("Processed Sales:", processed_sales)
    print("Processed Purchases:", processed_purchases)

if __name__ == "__main__":
    main()

File: examples/example6/dataProcessor/DataProcessor.java
package example6.dataProcessor;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

class DataProcessor {
    public List<Map<String, Object>> processSalesData(List<Map<String, Object>> salesData) {
        // Duplizierter Validierungscode
        if (salesData == null || salesData.isEmpty()) {
            throw new IllegalArgumentException("Sales data cannot be empty");
        }

        List<Map<String, Object>> processedSales = new ArrayList<>();
        for (Map<String, Object> sale : salesData) {
            // Duplizierte Validierungslogik
            if (!sale.containsKey("amount")) {
                throw new IllegalArgumentException("Invalid sale record: " + sale);
            }

            if (!sale.containsKey("date")) {
                throw new IllegalArgumentException("Invalid sale record: " + sale);
            }

            // Komplexe Verarbeitungslogik
            Map<String, Object> processedSale = new HashMap<>();
            processedSale.put("amount", Double.parseDouble(sale.get("amount").toString()));
            processedSale.put("date", normalizeDate(sale.get("date")));
            processedSale.put("is_valid", validateSale(sale));

            processedSales.add(processedSale);
        }

        return processedSales;
    }

    public List<Map<String, Object>> processPurchaseData(List<Map<String, Object>> purchaseData) {
        // Fast identischer Code wie bei Sales Data
        if (purchaseData == null || purchaseData.isEmpty()) {
            throw new IllegalArgumentException("Purchase data cannot be empty");
        }

        List<Map<String, Object>> processedPurchases = new ArrayList<>();
        for (Map<String, Object> purchase : purchaseData) {
            // Fast gleiche Validierungslogik
            if (!purchase.containsKey("amount")) {
                throw new IllegalArgumentException("Invalid purchase record: " + purchase);
            }

            if (!purchase.containsKey("date")) {
                throw new IllegalArgumentException("Invalid purchase record: " + purchase);
            }

            // Ähnliche Verarbeitungslogik
            Map<String, Object> processedPurchase = new HashMap<>();
            processedPurchase.put("amount", Double.parseDouble(purchase.get("amount").toString()));
            processedPurchase.put("date", normalizeDate(purchase.get("date")));
            processedPurchase.put("is_valid", validatePurchase(purchase));

            processedPurchases.add(processedPurchase);
        }

        return processedPurchases;
    }

    private String normalizeDate(Object date) {
        // Datumsformatierungslogik
        try {
            return date.toString();
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid date format: " + date);
        }
    }

    private boolean validateSale(Map<String, Object> sale) {
        // Verkaufsvalidierung
        return sale.containsKey("amount") &&
                sale.containsKey("date") &&
                Double.parseDouble(sale.get("amount").toString()) > 0;
    }

    private boolean validatePurchase(Map<String, Object> purchase) {
        // Kaufvalidierung
        return purchase.containsKey("amount") &&
                purchase.containsKey("date") &&
                Double.parseDouble(purchase.get("amount").toString()) > 0;
    }

    public static void main(String[] args) {
        DataProcessor processor = new DataProcessor();

        List<Map<String, Object>> salesData = new ArrayList<>();
        Map<String, Object> sale1 = new HashMap<>();
        sale1.put("amount", 100);
        sale1.put("date", "2023-01-01");
        salesData.add(sale1);

        List<Map<String, Object>> purchaseData = new ArrayList<>();
        Map<String, Object> purchase1 = new HashMap<>();
        purchase1.put("amount", 150);
        purchase1.put("date", "2023-03-01");
        purchaseData.add(purchase1);

        List<Map<String, Object>> processedSales = processor.processSalesData(salesData);
        List<Map<String, Object>> processedPurchases = processor.processPurchaseData(purchaseData);

        System.out.println("Processed Sales: " + processedSales);
        System.out.println("Processed Purchases: " + processedPurchases);
    }
}
